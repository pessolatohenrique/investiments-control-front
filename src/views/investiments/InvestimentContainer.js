import React, { useEffect, useState } from "react";
import axios from "axios";
import { useParams } from "react-router-dom";
import {
  Container,
  Box,
  Card,
  CardContent,
  Typography,
  Grid,
  TextField,
  Button,
  Snackbar,
  Alert,
  FormControl,
  InputLabel,
  Select,
  MenuItem,
  FormHelperText,
} from "@mui/material";

import {
  THEME_COLOR,
  SNACKBAR_DIRECTION,
} from "../../constants/default_settings";

import {
  MODAL_CONFIRM_TITLE,
  MODAL_CONFIRM_SUBTITLE,
  MODAL_CONFIRM_SUBTITLE_REDEEMED,
  MODAL_CONFIRM_SUBTITLE_CANCEL,
} from "../../constants/messages";

import BreadcrumbsWrapper from "../../components/BreadcrumbsWrapper";
import ModalWrapper from "../../components/ModalWrapper";

import useToast from "../../hooks/useToast";
import { InvestimentList } from "./InvestimentList";
import { orderByListDate } from "../../utils/orders";

function InvestimentContainer() {
  // custom hooks
  const { open, error, setError, showToast, hideToast } = useToast();

  const [investiments, setInvestiments] = useState([]);
  const [investimentsRedeemed, setInvestimentsRedeemed] = useState([]);
  const [investimentsByDate, setInvestimentsByDate] = useState([]);

  // modal
  const [showModalInvestiment, setShowModalInvestiment] = useState(false);
  const [selectedIdInvestiment, setSelectedIdInvestiment] = useState("");
  const [
    showModalInvestimentRedeemed,
    setShowModalInvestimentRedeemed,
  ] = useState(false);

  const [
    selectedIdInvestimentRedeemed,
    setSelectedIdInvestimentRedeemed,
  ] = useState("");

  const [showModalInvestimentCancel, setShowModalInvestimentCancel] = useState(
    false
  );
  const [
    selectedIdInvestimentCancel,
    setSelectedIdInvestimentCancel,
  ] = useState("");

  async function getInvestiments() {
    try {
      const response = await axios.get(`/investiment`);
      const investimentsResponse = response?.data.filter(
        (item) => !item.has_redeemed
      );
      const investimentsRedeemedResponse = [...response?.data].filter(
        (item) => item.has_redeemed
      );

      const investimentsOrderDate = orderByListDate(
        investimentsResponse,
        "final_date",
        "ASC"
      ).filter((item) => !item.has_redeemed && item.final_date !== null);

      setInvestiments(investimentsResponse);
      setInvestimentsRedeemed(investimentsRedeemedResponse);
      setInvestimentsByDate(investimentsOrderDate.slice(0, 9));
    } catch (error) {
      showToast();
      setError(error?.response?.data?.message || null);
    }
  }

  useEffect(() => {
    getInvestiments();
  }, []);

  return (
    <>
      <Box
        minHeight="100vh"
        sx={{
          backgroundColor: `#E5E5E5`,
          overflow: "hidden",
        }}
      >
        <Snackbar
          anchorOrigin={SNACKBAR_DIRECTION}
          open={open}
          autoHideDuration={6000}
          onClose={hideToast}
        >
          <Alert onClose={hideToast} severity="error" sx={{ width: "100%" }}>
            {error}
          </Alert>
        </Snackbar>

        {showModalInvestiment && (
          <ModalWrapper
            isOpen={showModalInvestiment}
            title={MODAL_CONFIRM_TITLE}
            subtitle={MODAL_CONFIRM_SUBTITLE}
            hasConfirmButton
            handleClose={() => {
              setSelectedIdInvestiment("");
              setShowModalInvestiment(false);
            }}
            endpoint={{
              method: "delete",
              name: `/investiment/${selectedIdInvestiment}`,
            }}
            callbackMethod={() => getInvestiments()}
          />
        )}

        {showModalInvestimentRedeemed && (
          <ModalWrapper
            isOpen={showModalInvestimentRedeemed}
            title={MODAL_CONFIRM_TITLE}
            subtitle={MODAL_CONFIRM_SUBTITLE_REDEEMED}
            hasConfirmButton
            confirmVariant="success"
            confirmMessage="Resgatar"
            handleClose={() => {
              setSelectedIdInvestimentRedeemed("");
              setShowModalInvestimentRedeemed(false);
            }}
            endpoint={{
              method: "put",
              name: `/investiment/${selectedIdInvestimentRedeemed}/redeemed`,
            }}
            callbackMethod={() => getInvestiments()}
          />
        )}

        {showModalInvestimentCancel && (
          <ModalWrapper
            isOpen={showModalInvestimentCancel}
            title={MODAL_CONFIRM_TITLE}
            subtitle={MODAL_CONFIRM_SUBTITLE_CANCEL}
            hasConfirmButton
            confirmMessage="Cancelar resgate"
            handleClose={() => {
              setSelectedIdInvestimentCancel("");
              setShowModalInvestimentCancel(false);
            }}
            endpoint={{
              method: "put",
              name: `/investiment/${selectedIdInvestimentCancel}/cancel`,
            }}
            callbackMethod={() => getInvestiments()}
          />
        )}

        <Grid
          container
          columnSpacing={3}
          marginLeft={2}
          marginRight={2}
          marginTop={5}
          marginBottom={3}
          sx={{ width: "96%" }}
        >
          <Grid item lg={12} md={12} sm={12} xs={11}>
            <Card>
              <CardContent>
                <BreadcrumbsWrapper
                  parentLink={{ link: "/", label: "Dashboard" }}
                  childrenLabel={"Investimentos"}
                />
                <Typography
                  gutterBottom
                  variant="h5"
                  component="h1"
                  color={THEME_COLOR}
                >
                  Investimentos pr√≥ximos do resgate
                </Typography>

                <InvestimentList
                  investiments={investimentsByDate}
                  onShowModal={() => setShowModalInvestiment(true)}
                  onSetSelectedId={(id) => setSelectedIdInvestiment(id)}
                  onShowModalRedeemed={() =>
                    setShowModalInvestimentRedeemed(true)
                  }
                  onSetSelectedIdRedeemed={(id) =>
                    setSelectedIdInvestimentRedeemed(id)
                  }
                />
              </CardContent>
            </Card>
          </Grid>
        </Grid>

        {/* <br /> */}
        <Grid
          container
          columnSpacing={3}
          marginLeft={2}
          marginRight={2}
          marginTop={5}
          marginBottom={3}
          sx={{ width: "96%" }}
        >
          <Grid item lg={12} md={12} sm={12} xs={11}>
            <Card>
              <CardContent>
                <BreadcrumbsWrapper
                  parentLink={{ link: "/", label: "Dashboard" }}
                  childrenLabel={"Investimentos"}
                />
                <Typography
                  gutterBottom
                  variant="h5"
                  component="h1"
                  color={THEME_COLOR}
                >
                  Investimentos resgatados
                </Typography>

                <InvestimentList
                  investiments={investimentsRedeemed}
                  onShowModal={() => setShowModalInvestiment(true)}
                  onSetSelectedId={(id) => setSelectedIdInvestiment(id)}
                  onShowModalCancel={() => setShowModalInvestimentCancel(true)}
                  onSetSelectedIdCancel={(id) =>
                    setSelectedIdInvestimentCancel(id)
                  }
                />
              </CardContent>
            </Card>
          </Grid>
        </Grid>

        <Grid
          container
          columnSpacing={3}
          marginLeft={2}
          marginRight={2}
          marginTop={5}
          marginBottom={3}
          sx={{ width: "96%" }}
        >
          <Grid item lg={12} md={12} sm={12} xs={11}>
            <Card>
              <CardContent>
                <BreadcrumbsWrapper
                  parentLink={{ link: "/", label: "Dashboard" }}
                  childrenLabel={"Investimentos"}
                />
                <Typography
                  gutterBottom
                  variant="h5"
                  component="h1"
                  color={THEME_COLOR}
                >
                  Investimentos
                </Typography>

                <InvestimentList
                  investiments={investiments}
                  onShowModal={() => setShowModalInvestiment(true)}
                  onSetSelectedId={(id) => setSelectedIdInvestiment(id)}
                  onShowModalRedeemed={() =>
                    setShowModalInvestimentRedeemed(true)
                  }
                  onSetSelectedIdRedeemed={(id) =>
                    setSelectedIdInvestimentRedeemed(id)
                  }
                />
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      </Box>
    </>
  );
}

export default InvestimentContainer;
