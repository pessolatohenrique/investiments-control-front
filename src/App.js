import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { isPageWithoutLogin } from "./utils/pages";
import useToken from "./hooks/useToken";
import AppBarWrapper from "./components/AppBarWrapper";
import NotFound from "./components/NotFound";
import LoginForm from "./views/register/LoginForm";
import RegisterContainer from "./views/register/RegisterContainer";
import DashboardContainer from "./views/dashboard/DashboardContainer";
import RecipeContainer from "./views/recipe/RecipeContainer";
import InvestimentContainer from "./views/investiments/InvestimentContainer";
import EarningForm from "./views/earnings/EarningForm";
import ExpenseForm from "./views/expenses/ExpenseForm";
import InvestimentForm from "./views/investiments/InvestimentForm";

function App(props) {
  const { token, setToken, setRefreshToken } = useToken();

  if (!token && !isPageWithoutLogin()) {
    return <LoginForm setToken={setToken} setRefreshToken={setRefreshToken} />;
  }

  return (
    <Router>
      <AppBarWrapper />

      <Switch>
        <Route exact path="/">
          <DashboardContainer />
        </Route>

        <Route exact path="/investimentos">
          <InvestimentContainer />
        </Route>

        <Route exact path="/investimento/novo">
          <InvestimentForm />
        </Route>

        <Route exact path="/investimento/:id">
          <InvestimentForm />
        </Route>

        <Route exact path="/receita/associar">
          <RecipeContainer />
        </Route>

        <Route exact path="/usuario/novo">
          <RegisterContainer />
        </Route>

        <Route exact path="/ganho/novo">
          <EarningForm />
        </Route>

        <Route exact path="/ganho/:id">
          <EarningForm />
        </Route>

        <Route exact path="/gasto/novo">
          <ExpenseForm />
        </Route>

        <Route exact path="/gasto/:id">
          <ExpenseForm />
        </Route>

        <Route path="*">
          <NotFound />
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
