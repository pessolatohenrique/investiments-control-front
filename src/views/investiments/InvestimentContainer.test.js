import axios from "axios";
import {
  render,
  screen,
  fireEvent,
  waitForElement,
} from "@testing-library/react";
import App from "../../App";
import InvestimentContainer from "./InvestimentContainer";

jest.mock("axios");

function mockCallValid() {
  axios.get.mockResolvedValueOnce({
    data: [
      {
        _id: "6375618ec9e12ca82d98afa7",
        userId: "630ff7a767ae1e8f6a17631b",
        type: "CHECKING_ACCOUNT",
        category: "Conta corrente",
        description: "Conta corrente",
        platform: "ItaÃº",
        net_value: 5000,
        __v: 0,
        profit: 0,
      },
      {
        _id: "637561adc9e12ca82d98afaa",
        userId: "630ff7a767ae1e8f6a17631b",
        goalId: "63507dcba3f4aa14eb41e0ec",
        type: "FIXED_INCOME",
        category: "Tesouro Direto",
        description: "IPCA+ 2026",
        indexer: {
          name: "IPCA",
          contracted_rate: 5.59,
          _id: "63979a5cb3871063914d5c33",
        },
        platform: "Clear",
        final_date: "2022-12-13T00:00:00.000Z",
        monthly_profitability: 0.95,
        invested_amount: 3000,
        expected_net_value: 5000,
        __v: 0,
        net_value: null,
        has_redeemed: false,
        expected_profit: 2000,
      },
    ],
  });
}

describe("Investiment Container", () => {
  beforeAll(() => {
    render(<App />);
    const submitBtn = screen.getByTestId("submit-button");
    const username = screen.getByTestId("username");
    const password = screen.getByTestId("password");

    fireEvent.input(username, {
      target: { value: process.env.REACT_APP_API_DEFAULT_USER },
    });
    fireEvent.input(password, {
      target: {
        value: process.env.REACT_APP_API_DEFAULT_EMAIL,
      },
    });
    fireEvent.click(submitBtn);
  });

  it("should generate snapshot", () => {
    const { container } = render(<InvestimentContainer />);
    expect(container).toMatchSnapshot();
  });

  it("should render investiments after request", async () => {
    mockCallValid();
    render(<InvestimentContainer />);

    await waitForElement(() => screen.getAllByText("Investimentos"));
  });
});
