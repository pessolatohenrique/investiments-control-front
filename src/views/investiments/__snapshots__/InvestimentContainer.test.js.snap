// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investiment Container should generate snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-m4l5cg"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-13puk6r-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 css-25posd-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <nav
              aria-label="breadcrumb"
              class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
            >
              <ol
                class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
              >
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                    data-testid="bread-firstlevel"
                    href="/"
                  >
                    Início
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
                >
                  ›
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                    data-testid="bread-secondlevel"
                    href="/"
                  >
                    Dashboard
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
                >
                  ›
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
                    data-testid="bread-lastlevel"
                  >
                    Investimentos
                  </p>
                </li>
              </ol>
            </nav>
            <h1
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1y15g74-MuiTypography-root"
            >
              Investimentos próximos do resgate
            </h1>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-13puk6r-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 css-25posd-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <nav
              aria-label="breadcrumb"
              class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
            >
              <ol
                class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
              >
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                    data-testid="bread-firstlevel"
                    href="/"
                  >
                    Início
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
                >
                  ›
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                    data-testid="bread-secondlevel"
                    href="/"
                  >
                    Dashboard
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
                >
                  ›
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
                    data-testid="bread-lastlevel"
                  >
                    Investimentos
                  </p>
                </li>
              </ol>
            </nav>
            <h1
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1y15g74-MuiTypography-root"
            >
              Investimentos resgatados
            </h1>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-13puk6r-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 css-25posd-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <nav
              aria-label="breadcrumb"
              class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
            >
              <ol
                class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
              >
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                    data-testid="bread-firstlevel"
                    href="/"
                  >
                    Início
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
                >
                  ›
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-9smqfy-MuiTypography-root-MuiLink-root"
                    data-testid="bread-secondlevel"
                    href="/"
                  >
                    Dashboard
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
                >
                  ›
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
                    data-testid="bread-lastlevel"
                  >
                    Investimentos
                  </p>
                </li>
              </ol>
            </nav>
            <h1
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1y15g74-MuiTypography-root"
            >
              Investimentos
            </h1>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
